astor/code_gen.py

```diff
@@ -638,22 +639,29 @@ def visit_Num(self, node,
                   # constants
                   new=sys.version_info >= (3, 0)):
         with self.delimit(node) as delimiters:
-            s = repr(node.n)
-
-            # Deal with infinities -- if detected, we can
-            # generate them with 1e1000.
-            signed = s.startswith('-')
-            if s[signed].isalpha():
-                im = s[-1] == 'j' and 'j' or ''
-                assert s[signed:signed + 3] == 'inf', s
-                s = '%s1e1000%s' % ('-' if signed else '', im)
-            self.write(s)
+            if cmath.isinf(node.n) or cmath.isnan(node.n):
+                # Represent infinity as 1e1000 and NaN as 1e1000-1e1000.
+                x = complex(node.n)
+                real, imag = (
```

Comment:
I'm not sure what would be readable enough for your tastes but also fully solve the problem. I know the code is ugly, but the problem is ugly, because of all the cases that need to be considered (finite vs. infintie vs NaN, real type vs. complex type with 0 imaginary part vs. complex type with 0 real part vs. complex type with nonzero parts). I'm guessing you wouldn't think it an improvement if I just replaced all the `if` expressions with `if` statements, which would make the code longer but probably not any easier to digest.

Reply:
Actually, writing that description of the cases gives me an idea. I'll try that now.

Reply:
I don't think the code is ugly, but I find following all the chained `a if b and c else d if ...` expressions a bit hard. Replacing them with if statements would make future maintenance of this code easier (and that also means I won't have disturb you every time I touch it :))

Reply:
Okay, check out the second commit.

Reply:
Looks great, thanks!